hexrd/transforms/cpp_sublibrary/src/transforms.cpp

```diff
@@ -0,0 +1,401 @@
+const double sqrtEps = std::sqrt(std::numeric_limits<double>::epsilon());
+
+
+const static Eigen::MatrixXd makeBinaryRotMat(const Eigen::Vector3d& a) {
+  Eigen::Matrix3d r = 2.0 * a * a.transpose();
+  r.diagonal() -= Eigen::Vector3d::Ones();
+  return r;
+}
+
+Eigen::MatrixXd makeRotMatOfExpMap(const Eigen::Vector3d& e) {
+  Eigen::AngleAxisd rotation(e.norm(), e.normalized());
+  return rotation.toRotationMatrix();
+}
+
+static Eigen::MatrixX3d makeOscillRotMat(const double chi, const Eigen::VectorXd& ome) {
+  Eigen::MatrixX3d rots(3 * ome.size(), 3);
+  const double cchi = cos(chi), schi = sin(chi);
+  xsimd::batch<double> come_v, some_v;
+
+  const size_t batch_size = xsimd::simd_traits<double>::size;
+  const size_t vec_size = ome.size();
+
+  double come_arr[batch_size], some_arr[batch_size];
+
+  for (size_t i = 0; i < vec_size; i += batch_size) {
+    auto ome_v = xsimd::load_unaligned(&ome[i]);
+    std::tie(some_v, come_v) = xsimd::sincos(ome_v);
+    come_v.store_aligned(come_arr);
+    some_v.store_aligned(some_arr);
+
+    for(size_t j = 0; j < batch_size && (i+j) < vec_size; ++j) {
+      rots.block<3, 3>(3*(i+j), 0) <<         come_arr[j],    0,         some_arr[j],
+                                       schi * some_arr[j], cchi, -schi * come_arr[j],
+                                      -cchi * some_arr[j], schi,  cchi * come_arr[j];
+    }
+  }
+  return rots;
+}
+
+// Both anglesToGvec and anglesToDvec, depending on if the last argument is true or false
+const static Eigen::MatrixX3d anglesToVec(const Eigen::MatrixXd &angs,
+                                    const Eigen::Vector3d &bHat_l,
+                                    const Eigen::Vector3d &eHat_l, double chi,
+                                    const Eigen::Matrix3d &rMat_c, bool toGVec) noexcept {
+  constexpr size_t batch_size = xsimd::simd_traits<double>::size;
+  const size_t vec_size = angs.rows();
+  const size_t num_full_batches = vec_size / batch_size;
+
+  auto cchi = cos(chi);
+  auto schi = sin(chi);
+  Eigen::MatrixX3d gVec_c(angs.rows(), 3);
+
+  const Eigen::Vector3d yHat = eHat_l.cross(bHat_l).normalized();
+  const Eigen::Vector3d bHat = bHat_l.normalized();
+  const Eigen::Vector3d xHat = bHat.cross(yHat);
+  Eigen::Matrix3d rotMat;
+  rotMat << xHat, yHat, -bHat;
+
+  auto rotMat00 = rotMat(0, 0);
+  auto rotMat01 = rotMat(0, 1);
+  auto rotMat02 = rotMat(0, 2);
+  auto rotMat10 = rotMat(1, 0);
+  auto rotMat11 = rotMat(1, 1);
+  auto rotMat12 = rotMat(1, 2);
+  auto rotMat20 = rotMat(2, 0);
+  auto rotMat21 = rotMat(2, 1);
+  auto rotMat22 = rotMat(2, 2);
+
+  auto mat00 = rMat_c(0, 0);
+  auto mat01 = rMat_c(0, 1);
+  auto mat02 = rMat_c(0, 2);
+  auto mat10 = rMat_c(1, 0);
+  auto mat11 = rMat_c(1, 1);
+  auto mat12 = rMat_c(1, 2);
+  auto mat20 = rMat_c(2, 0);
+  auto mat21 = rMat_c(2, 1);
+  auto mat22 = rMat_c(2, 2);
+
+  // SIMD loop for full batches
+  for (size_t i = 0; i < num_full_batches * batch_size; i += batch_size) {
```

Comment:
Can we use xsimd::transform here to avoid writing this loop body twice?

Reply:
Lets not bother with this. Apparently xsimd::transform was removed recently and is currently in limbo.
https://github.com/xtensor-stack/xsimd/issues/1045